diff --git a/binding.gyp b/binding.gyp
index f21c7df48b17e5c173446fb722e1229095282bde..0860e739ce4d8734515ce14e6c5d7ca4c733f6ea 100644
--- a/binding.gyp
+++ b/binding.gyp
@@ -6,7 +6,7 @@
     'conditions': [
       ['OS=="win"', {
         'msvs_configuration_attributes': {
-          'SpectreMitigation': 'Spectre'
+          'SpectreMitigation': 'false'
         },
         'msvs_settings': {
             'VCCLCompilerTool': {
diff --git a/deps/winpty/src/winpty.gyp b/deps/winpty/src/winpty.gyp
index 1ac5758bedd8cf54f32280dea4e4aeb5afdee30d..d2a5e7911fb823bef3c9d57d2bf35830f8cde6ef 100644
--- a/deps/winpty/src/winpty.gyp
+++ b/deps/winpty/src/winpty.gyp
@@ -10,7 +10,7 @@
     #    make -j4 CXX=i686-w64-mingw32-g++ LDFLAGS="-static -static-libgcc -static-libstdc++"
 
     'variables': {
-        'WINPTY_COMMIT_HASH%': '<!(cmd /c "cd shared && GetCommitHash.bat")',
+        'WINPTY_COMMIT_HASH%': '<!(cmd /c "cd shared && .\\GetCommitHash.bat")',
     },
     'target_defaults' : {
         'defines' : [
@@ -22,7 +22,7 @@
         'include_dirs': [
             # Add the 'src/gen' directory to the include path and force gyp to
             # run the script (re)generating the version header.
-            '<!(cmd /c "cd shared && UpdateGenVersion.bat <(WINPTY_COMMIT_HASH)")',
+            '<!(cmd /c "cd shared && .\\UpdateGenVersion.bat <(WINPTY_COMMIT_HASH)")',
         ]
     },
     'targets' : [
@@ -41,7 +41,7 @@
                 '-luser32',
             ],
             'msvs_configuration_attributes': {
-                'SpectreMitigation': 'Spectre'
+                'SpectreMitigation': 'false'
             },
             'msvs_settings': {
                 # Specify this setting here to override a setting from somewhere
@@ -143,7 +143,7 @@
                 '-luser32',
             ],
             'msvs_configuration_attributes': {
-                'SpectreMitigation': 'Spectre'
+                'SpectreMitigation': 'false'
             },
             'msvs_settings': {
                 # Specify this setting here to override a setting from somewhere
