diff --git a/lib/connection.js b/lib/connection.js
index 9e24391b6bdcfd8677a6d0dd51481c1231990159..058109c8c0f04980f45402cdb2e3a6ca5a981ee9 100644
--- a/lib/connection.js
+++ b/lib/connection.js
@@ -83,19 +83,18 @@ class Connection extends EventEmitter {
       const options = {
         socket: self.stream,
       }
-
-      if (self.ssl !== true) {
-        Object.assign(options, self.ssl)
-
-        if ('key' in self.ssl) {
-          options.key = self.ssl.key
-        }
-      }
-
-      if (net.isIP(host) === 0) {
-        options.servername = host
-      }
       try {
+        if (self.ssl !== true) {
+          Object.assign(options, self.ssl)
+  
+          if ('key' in self.ssl) {
+            options.key = self.ssl.key
+          }
+        }
+  
+        if (net.isIP(host) === 0) {
+          options.servername = host
+        }
         self.stream = tls.connect(options)
       } catch (err) {
         return self.emit('error', err)